### ISSUE TITLE
Lint to check for unwrap except on Infallible error

### ISSUE THREAD

**Issue by conqp** [rust-lang/rust-clippy#11245]
### What it does

Behaves exactly like `unwrap_used` _**except**_ if the error type is `Infallible` it does not issue a warning.

### Advantage

- Still warns the user about unsafe unwraps.
- Does not bother the user where `unwrap` cannot fail due to the error being `Infallible`.

### Drawbacks

None (that I can think of)

### Example

```rust
const fn foo() -> Result<&'static str, Infallible> {
    Ok("foo")
}

fn main() {
    let f = foo().expect("This cannot fail.");
    println!("{f}");
}
```

Could be written as:

```rust
const fn foo() -> Result<&'static str, Infallible> {
    Ok("foo")
}

fn main() {
    let f = foo().unwrap();
    println!("{f}");
}
```

**Comment by conqp** [comment]
I don't know what the policy of this project regarding backwards compatibility is, but it may also be an option to change the behaviour of `unwrap_used` to not just issue a warning in this case instead of introducing a new lint.

**Comment by Centri3** [comment]
It also seems to lint on `!` as well. We shouldn't for never and zero-variant enums as they can never be constructed (thus will never panic)
